internal/services/loganalytics/log_analytics_solution_resource.go

```diff
@@ -4,255 +4,311 @@
+					state.SolutionName = solutionName
+					state.WorkspaceName = workspaceName
+				}
 
-			solutionName := segments[0]
-			workspaceName := strings.TrimSuffix(segments[1], ")")
-			d.Set("solution_name", solutionName)
-			d.Set("workspace_name", workspaceName)
-		}
-
-		if props := model.Properties; props != nil {
-			var workspaceId string
-			if props.WorkspaceResourceId != "" {
-				id, err := workspaces.ParseWorkspaceIDInsensitively(props.WorkspaceResourceId)
-				if err != nil {
-					return err
+				if props := model.Properties; props != nil {
+					var workspaceId string
+					if props.WorkspaceResourceId != "" {
+						id, err := workspaces.ParseWorkspaceIDInsensitively(props.WorkspaceResourceId)
+						if err != nil {
+							return err
+						}
+						workspaceId = id.ID()
+					}
+					state.WorkspaceResourceId = workspaceId
+				}
+
+				if plan := model.Plan; plan != nil {
+					state.SolutionPlan = flattenAzureRmLogAnalyticsSolutionPlan(plan)
 				}
-				workspaceId = id.ID()
-			}
-			d.Set("workspace_resource_id", workspaceId)
-		}
 
-		if err := d.Set("plan", flattenAzureRmLogAnalyticsSolutionPlan(model.Plan)); err != nil {
-			return fmt.Errorf("setting `plan`: %+v", err)
-		}
+				state.Tags = pointer.From(model.Tags)
 
-		if err = tags.FlattenAndSet(d, flattenTags(model.Tags)); err != nil {
-			return err
-		}
+			}
+
+			return metadata.Encode(&state)
+		},
 	}
-	return nil
 }
 
-func resourceLogAnalyticsSolutionDelete(d *pluginsdk.ResourceData, meta interface{}) error {
-	client := meta.(*clients.Client).LogAnalytics.SolutionsClient
-	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
-	defer cancel()
-	id, err := solution.ParseSolutionID(d.Id())
-	if err != nil {
-		return err
-	}
+func (s LogAnalyticsSolutionResource) Delete() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.LogAnalytics.SolutionsClient
+
+			id, err := solution.ParseSolutionID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			err = client.DeleteThenPoll(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("deleting %s: %+v", *id, err)
+			}
 
-	err = client.DeleteThenPoll(ctx, *id)
-	if err != nil {
-		return fmt.Errorf("deleting %s: %+v", *id, err)
+			return nil
+		},
 	}
+}
+
+func (s LogAnalyticsSolutionResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.LogAnalytics.SolutionsClient
+
+			id, err := solution.ParseSolutionID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			var config SolutionResourceModel
+			if err := metadata.Decode(&config); err != nil {
+				return fmt.Errorf("decoding %v", err)
+			}
 
-	return nil
+			payload := solution.SolutionPatch{
```

Comment:
We should use `metadata.ResourceData.HasChange` on tags here.

Reply:
`tag` is the only property could be changed, I'm considering `HasChange()` will alwasy return `true`? or the `update` func will not be called.

Reply:
We still need to call `HasChange` on the property for `ignore_changes` to work on it.

Reply:
got it